shared/js/content-scope/fingerprint.js

```diff
@@ -0,0 +1,111 @@
+/* global sjcl */
+function getCanvasKeySync (sessionKey, domainKey, inputData) {
+    // eslint-disable-next-line new-cap
+    let hmac = new sjcl.misc.hmac(sjcl.codec.utf8String.toBits(sessionKey + domainKey), sjcl.hash.sha256)
+    return sjcl.codec.hex.fromBits(hmac.encrypt(inputData))
+}
+
+// linear feedback shift register to find a random approximation
+function nextRandom (v) {
+    return Math.abs((v >> 1) | (((v << 62) ^ (v << 61)) & (~(~0 << 63) << 62)))
+}
+
+let exemptionList = []
+
+function shouldExemptUrl (url) {
+    for (let regex of exemptionList) {
+        if (regex.test(url)) {
+            return true
+        }
+    }
+    return false
+}
+
+function initExemptionList (stringExemptionList) {
+    for (let stringExemption of stringExemptionList) {
+        exemptionList.push(new RegExp(stringExemption))
+    }
+}
+
+// Checks the stack trace if there are known libraries that are broken.
+function shouldExemptMethod () {
+    let errorLines = new Error().stack.split('\n')
+    let errorFiles = new Set()
+    // Should cater for Chrome and Firefox stacks, we only care about https? resources.
+    let lineTest = /(\()?(http[^)]+):[0-9]+:[0-9]+(\))?/
+    for (let line of errorLines) {
+        let res = line.match(lineTest)
+        if (res) {
+            let path = res[2]
+            // checked already
+            if (errorFiles.has(path)) {
+                continue
+            }
+            if (shouldExemptUrl(path)) {
+                console.log('Exempting script path:', path)
```

Comment:
We should put logging behind a debug flag before release to avoid polluting page logs.

Reply:
This is the status quo, I do agree though.

Reply:
Maybe we should remove all logging calls from the scripts that run on page? Other ones might have been left in by mistake, unless they are informative to a user debugging for themselves I suppose.